### ISSUE TITLE
Coding complexity sniff generates different numbers to the "official" implementation.

### ISSUE THREAD

**Issue by cooperaj** [slevomat/coding-standard#1535]
The Sonarlint addon for PHPStorm runs over the following method and generates a complexity of "4". The same function is using the complexity sniff is found to have a complexity of "7". This is over the default warn value of 5 and so a warning is generated (and my CI fails)

The Sonarlint addon uses a default complexity of 15. 

Given the algorithm is derived from Sonarcubes whitepaper on the problem I think something might not be correct. This is aside from the default warn value being (imo) unreasonably low. 

```
public function handle(ServerRequestInterface $request): ResponseInterface
    {
        $requestData = $request->getParsedBody();
        $userId      = $request->getHeader('user-token')[0];

        if (
            empty($requestData['reference_number']) ||
            empty($requestData['dob']) ||
            empty($requestData['first_names']) ||
            empty($requestData['last_name']) ||
            empty($requestData['postcode'])
        ) {
            throw new BadRequestException('Required data missing to request an activation key');
        }

        $lpaMatchResponse = $this->addAccessForAllLpa->validateRequest($userId, $requestData);

        if (($this->featureEnabled)('allow_older_lpas')) {
            ($this->checkLpaCleansed)($userId, $lpaMatchResponse);
        }

        $this->accessForAllLpaService->requestAccessByLetter(
            (string) $requestData['reference_number'],
            $lpaMatchResponse['actor']['uId'],
            $userId,
            $lpaMatchResponse['lpaActorToken'] ?? null,
        );

        $this->logger->notice(
            'Successfully matched data for LPA {uId} and requested letter for account {id} ',
            [
                'event_code' => $lpaMatchResponse['caseSubtype'] === 'hw'
                    ? EventCodes::FULL_MATCH_KEY_REQUEST_SUCCESS_LPA_TYPE_HW
                    : EventCodes::FULL_MATCH_KEY_REQUEST_SUCCESS_LPA_TYPE_PFA,
                'id'         => $userId,
                'uId'        => (string)$requestData['reference_number'],
                'lpaType'    => $lpaMatchResponse['caseSubtype'],
            ],
        );

        return new EmptyResponse();
    }
```

**Comment by cooperaj** [comment]
The Sonarlint plugin is quite handy as it not only warns when things break the rules but it provides inline counts of the incrementing complexity number.

![Screenshot 2023-03-20 at 18 11 11](https://user-images.githubusercontent.com/400210/226429539-4b8674b4-1c89-4203-99a3-88e2f8de37df.png)

**Comment by bkdotcom** [comment]
The issue lies in implementing

> ### Sequences of logical operators
> For similar reasons, Cognitive Complexity does not increment for each binary
logical operator. Instead, it assesses a fundamental increment for each sequence
of binary logical operators.


> Because boolean expressions become more difficult to understand with mixed
operators, Cognitive complexity increments for each new sequence of like
operators. For instance:

```
if (a          // +1 for `if`
    && b && c  // +1
    || d || e  // +1
    && f)      // +1

if (a        // +1 for `if`
   &&        // +1
  !(b && c)) // +1
```

>While Cognitive Complexity offers a “discount” for like operators relative to
Cyclomatic Complexity, it does increment for all sequences of binary boolean
operators such as those in variable assignments, method invocations, and return
statements.

... non trivial

**Comment by bkdotcom** [comment]
I will add that it's not clear what constitutes a "new sequence" of operators

what does sonar say the complexity of this function is?

```php
    private function test4byteSeq()
    {
        $bytes = $this->getBytes(4);
        if (\count($bytes) !== 4) {
            return false;
        }
        return (($bytes[1] & 0xc0) !== 0x80
            || ($bytes[2] & 0xc0) !== 0x80
            || ($bytes[3] & 0xc0) !== 0x80
            || ($bytes[0] === 0xf0
                && ($bytes[1] & 0xf0) === 0x80)  // overlong
            || ($bytes[0] === 0xf4
                && $bytes[1] > 0x8f)
            || $bytes[0] > 0xf4    // > U+10FFFF
        ) === false;
    }
```
